### ISSUE TITLE
Digest encoding

### ISSUE THREAD

**Issue by pwinckles** [OCFL/spec#534]
It would be nice if the [digests section](https://ocfl.io/1.0/spec/#digests) specified outside of the digest table that digests MUST be hex encoded. As it is currently written, there are 4 different codes (29-32) that all specify that a specific digest algorithm must be hex encoded. Not, that there are 5 algorithms in the table, but the language for `md5` is different from the rest and has no associated MUST.

Ideally, there would be one code that says digests MUST be hex encoded with the language in the table simplified.

**Comment by zimeon** [comment]
What if we were to later add a new digest that used something other than hex encoding?

The md5 case is slightly different in that the original RFC describes the output as hex encoded rather than admitting multiple encoding possibilities, so I don't think an additional MUST is needed.

**Comment by pwinckles** [comment]
What is the error code if a md5 digest contains a non-hex character?

I see what you're saying. It would seem more like you're allowing for the possibility of an extension algorithm to be, say, base64 encoded. The spec itself only deals with hex encoding and I would assume that doing otherwise would require a new OCFL version, in which case the language would necessarily change.

**Comment by zimeon** [comment]
Good question re. what error -- the digests section doesn't cover the case of digests that are illegal according to the algorithm. My code gives an https://ocfl.io/1.0/spec/#E057 if a digest in fixity has the wrong form, https://ocfl.io/1.0/spec/#E025 is a digest in manifest has the wrong form

**Comment by pwinckles** [comment]
Isn't `E057` just saying that digest algorithm keys must contain a map of digest strings to content path arrays?

I feel like I am being too pedantic here, but I am really struggling with the dissonance in specificity of the validation codes.

**Comment by zimeon** [comment]
I think E057 is reasonable for any malformed digest because "a key corresponding to the digest value" can be true only if the digest value is a valid. Internally, I split E057 into 4 different cases to cover different aspects (https://github.com/zimeon/ocfl-py/blob/main/ocfl/data/validation-errors.json#L338-L360):
```
"E057a": {
    "params": ["where", "algorithm"],
    "description": {
       "en": "OCFL Object %s inventory fixity block entry for digest algorithm name %s is not a JSON object"
    }
  },
  "E057b": {
    "params": ["where", "algorithm", "digest"],
    "description": {
       "en": "OCFL Object %s inventory fixity block entry for digest algorithm %s includes digest %s which has the wrong form"
    }
  },
  "E057c": {
    "params": ["where", "algorithm", "digest"],
    "description": {
       "en": "OCFL Object %s inventory fixity block entry for digest algorithm %s, digest %s is not a JSON list"
    }
  },
  "E057d": {
    "params": ["where", "algorithm", "digest", "path"],
    "description": {
       "en": "OCFL Object %s inventory fixity block entry for digest algorithm %s, digest %s includes a content path %s that is not in the manifest"
    }
```

This whole process is quite interesting to me. This is the first time I'd tried to write code that links to specific MUSTs in the spec. It highlights the difference between the goals of writing instructions for doing something vs. instructions for validating something. This might be an interesting topic for a future community call.

**Comment by pwinckles** [comment]
I am up for discussing on the next call.

I appreciate the goal, and I agree that there does seem to be a difference in what makes for a well written spec and what makes for exhaustive validation codes.

From my perspective, the validation codes could either be a handful of general classes of validation concerns or an exhaustive list of every conceivable specific violation of the OCFL spec. The existing codes attempt to do the later by mapping MUSTs to codes, but fall short of full comprehension because not every constraint is signaled by a MUST in the spec (I do not think this is a bad thing).

The result is that MUSTs are in effect used tie sentences in the spec to codes that can be applied to any violation so long as a constraint can be inferred from the sentence. This is where the disconnect is for me. There are codes like `E077`-`E079` that call out 3 different MUSTs within the same sentence that seem like they should actually be just a single code, and codes like `E057` that have only a single MUST but can be inferred to many any number of things, depending on what other holes there are in the codes.

**Comment by pwinckles** [comment]
One more comment just so my thoughts are collected. I think why I find the above frustrating is that the codes give the appearance of being specific, but in a lot of cases they aren't, which requires me to spend a lot of time hunting through the codes trying to find one that might be reasonable to apply. This would not be the case if the codes were either more general or more specific.

Another example, if an object contains `v1` but it is a file and not a directory, what is the error code? Perhaps one of `E011`-`E014` that talk about version directory naming conventions? Or maybe it's just the most general `E001`?

**Comment by julianmorley** [comment]
> Another example, if an object contains v1 but it is a file and not a directory, what is the error code? Perhaps one of E011-E014 that talk about version directory naming conventions? Or maybe it's just the most general E001?

I think I'd go with E001 there; that's a file that's not meant to be in the base directory, and it doesn't matter that it happens to have the same string value as an allowable directory entry.

**Comment by zimeon** [comment]
I agree. I created an example https://github.com/OCFL/fixtures/pull/51 and it seems that I also agreed with `E001`:
```
fixtures> ocfl-validate.py 1.0/bad-objects/E001_v2_file_in_root
[E001a] OCFL Object root contains unexpected file: v2 (see https://ocfl.io/1.0/spec/#E001)
INFO:ocfl.object:OCFL object at 1.0/bad-objects/E001_v2_file_in_root is INVALID
```

**Comment by awoods** [comment]
This issue ([issue-534](https://github.com/OCFL/spec/issues/534)) is more of a discussion topic than a request/suggestion for specific change in the spec. I would suggest that we continue with the approach of trying to capture detailed requirements with specific `MUST` language, and raise points of ambiguity on a case by case basis... with the common result being more `MUST` language. Reverting validation codes to more general error categories would represent a loss at this point.

I suggest closing this issue, and potentially continuing the conversation on the next community call.

**Comment by rosy1280** [comment]
agenda created: https://github.com/OCFL/spec/wiki/2021.11.10-Community-Meeting, agenda item added.  closing this ticket.
